util/src/main/java/io/kubernetes/client/leaderelection/resourcelock/ResourceLock.java

```diff
@@ -0,0 +1,178 @@
+package io.kubernetes.client.leaderelection.resourcelock;
+
+import com.google.gson.JsonObject;
+import com.google.gson.reflect.TypeToken;
+import com.squareup.okhttp.Call;
+import io.kubernetes.client.ApiClient;
+import io.kubernetes.client.ApiException;
+import io.kubernetes.client.ApiResponse;
+import io.kubernetes.client.leaderelection.LeaderElectionRecord;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class ResourceLock implements Lock {
+    private static final String METADATA = "metadata";
+    private static final String ANNOTATIONS = "annotations";
+
+    private final String name;
+    private final String namespace;
+
+    private final ApiClient client;
+    private final String apiVersion;
+    private final String kind;
+    private final String type;
+
+    private final ResourceLockConfig resourceLockConfig;
+    private JsonObject object;
+
+    public ResourceLock(String name, String namespace, ApiClient client, String apiVersion, String kind, String type,
+                        ResourceLockConfig resourceLockConfig) {
+        this.name = name;
+        this.namespace = namespace;
+        this.client = client;
+        this.apiVersion = apiVersion;
+        this.kind = kind;
+        this.type = type;
+        this.resourceLockConfig = resourceLockConfig;
+    }
+
+    private String getRootUrl() {
+        return String.format("/api/%s/namespaces/%s/%s",
```

Comment:
we're introducing `coordination` api in the future releases, so it doesn't have to be a legacy`/api` path. dealing w/ format printing url here doesn't make sense. better make it an interface or lambda returning an (generic?) object

Reply:
Good idea, i will change this and add more unit test, it will take a time to complete

Reply:
there is a not good design in `k8s-java-client`, because all k8s generated resource type  do not have an  common parent class. if `ConfigMap` and `Endpoint` extends `HashedMetaObject`, then this interface will be more meaningful.

Reply:
>  if ConfigMap and Endpoint extends HashedMetaObject, then this interface will be more meaningful.

we have limitted kinds of resources for leader-electing. i presume just multiple independant resource lock classes is fine. separated class like `ConfigMapResourceLock`, `EndpointsResourceLock`, `CoordinationResourceLock`..

@brendandburns for thoughts